File: samples/client/petstore/java/src/main/java/org/openapijsonschematools/client/schemas/validation/TypeValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.checkerframework.checker.nullness.qual.Nullable;
import org.openapijsonschematools.client.exceptions.ValidationException;

import java.util.List;
import java.util.Map;

public class TypeValidator implements KeywordValidator {
    @Override
    public @Nullable PathToSchemasMap validate(
        ValidationData data
    ) throws ValidationException {
        var type = data.schema().type;
        if (type == null) {
            return null;
        }
        Class<?> argClass;
        var arg = data.arg();
        if (arg == null) {
            argClass = Void.class;
        } else if (arg instanceof List) {
            argClass = List.class;
        } else if (arg instanceof Map) {
            argClass = Map.class;
        } else {
            argClass = arg.getClass();
        }
        if (!type.contains(argClass)) {
            throw new ValidationException("invalid type");
        }
        return null;
    }
}

File: samples/client/petstore/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/TypeValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class TypeValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        if (data.schema.type == null) {
            return null
        }
        val argClass = when (data.arg) {
            null -> {
                Void::class.java
            }
            is List<*> -> {
                MutableList::class.java
            }

            is Map<*, *> -> {
                MutableMap::class.java
            }

            else -> {
                data.arg.javaClass
            }
        }
        if (!data.schema.type.contains(argClass)) {
            throw ValidationException("invalid type")
        }
        return null
    }
}
